### PR TITLE
Refactor resource builder utilities to support streams

### PR DESCRIPTION
**Pull request by 0xazure**
This is part one of a two-step refactoring to improve the build scripts from a maintenence point of view.

The primary change is to refactor the resource builder code to support streams and make them more like a traditional `gulp` plugin.  The resource builders now consume and return streams of files to fit into the `gulp.pipe()` paradigm.

This change is particularly necessary for part two of the build script refactoring (coming soon :tm:) that relies on the ability to compose streams to build up pipe chains programatically.

Changes include:
- The aforementioned refactor to support streams
- Moving the resource builder files out of `Extensions/dist/` and into `dev/builders` to separate them from our generated distribution files
- Major refactor to reduce code duplication and improve use of generalized functions
- Refactor to merge certain related functions and completely remove unused ones

### CODE REVIEW

**Comment by 0xazure** [comment]
My biggest concern is ensuring that the output of `gulp build:extensions` and `gulp build:themes` hasn't changed in any way that would break XKit parsing and loading, so that's primarily what needs to be tested and reviewed.

Also, please pay attention to the function documentation and let me know if it's lacking anywhere.  I did my best to describe the functions as best as I could, but a fresh set (or more) of eyes can always help improve it.


When I was digging into some of the resource object building logic, I got a little confused around some of the extension keys' value types.  Should gallery keys like `errors`, `frame`, `beta`, etc be strings or bools?  The currently-deployed extensions (ex. [accesskit.json](https://github.com/new-xkit/XKit/blob/86765698c690454577120c562cfbbf736a14bd01/Extensions/dist/accesskit.json)) have all key values as strings except for `error` which is a bool, but looking at what's deployed from old XKit (ex. [accesskit.json](http://puaga.com/seven/get.php?extension=accesskit)) everything is a string, including `errors`.

I've had a look at the [download parsing logic](https://github.com/new-xkit/XKit/blob/c8f9755ad15bc8cadccd90bbe552a1a7079927bf/xkit.js#L1147-L1271), and I'd like some verification from someone else as to what the types should be.  It _looks_ like [strings get converted](https://github.com/new-xkit/XKit/blob/c8f9755ad15bc8cadccd90bbe552a1a7079927bf/xkit.js#L1199), so everything should be a string?

**Comment by hobinjk** [comment]
Somewhere `errors` in the extension's json file becomes `"false"` instead of `false`, causing an error in `install_extension` as implemented in `xkit.js`. Apparently `"false"` is considered a truthy value by JavaScript.

File: dev/builders/extension.js [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+		 * {boolean} frame       - Value of the FRAME field in `script`
+		 * {boolean} [beta]      - Value of the optional BETA field in `script`
+		 * {boolean} [slow]      - Value of the optional SLOW field in `script`
+		 */
+		var extension = {};
+		extension.id = path.basename(file.path, '.js');
+		extension.script = file.contents.toString();
+		extension.file = 'found';
+		extension.server = 'up';
+		extension.errors = 'false';
```
Comment (hobinjk) [inline dev/builders/extension.js]: This is a truthy value, see previous comment

**Comment by 0xazure** [comment]
Looks like when we converted some `=== 'false'` into `===  false`, [this part of the parsing logic](https://github.com/new-xkit/XKit/blame/c8f9755ad15bc8cadccd90bbe552a1a7079927bf/xkit.js#L1235) didn't get similar conversions like [the `slow` field](https://github.com/new-xkit/XKit/blame/c8f9755ad15bc8cadccd90bbe552a1a7079927bf/xkit.js#L1219-L1223).

Where do we want to fix this: in `xkit.js` or in the JSON that gets built?  Changing the builder to use a boolean is easier, certainly.

**Comment by hobinjk** [comment]
Based on there still being people using 7.6.x, making the change in the JSON seems preferable

**Comment by 0xazure** [comment]
Updated `extension.errors` to use a boolean, as expected by `xkit.js`.

**Comment by hobinjk** [comment]
`set_setting` doesn't return the extension, it returns one of two objects (where errors is always a boolean)

**Comment by 0xazure** [comment]
I'm not actually sure `extension.errors` was actually the issue.  The [server response gets `JSON.parse()`'d ](https://github.com/new-xkit/XKit/blob/c8f9755ad15bc8cadccd90bbe552a1a7079927bf/xkit.js#L320), so it should still convert it into a bool.

**Comment by hobinjk** [comment]
```javascript
JSON.parse('{"a": "false"}').a
```
and
```javascript
JSON.parse('{"a": false}').a
```
are different.

I am unable to test on Chrome because the cert is not being accepted, but this looks good to me on Firefox.

**Comment by 0xazure** [comment]
You're completely correct (apparently I need a refresher on the ins and outs of `JSON.parse()`).  I'm just still mystified as to how it worked originally when the older servers serve a string.

I went and tested thoroughly on Chrome, can confirm it's working as expected.

**Comment by hobinjk** [comment]
Awesome! Glad to be able to merge in this great work.
@homu r+

**Pull request merged**